### ISSUE TITLE
Low recall rate on WIDER FACE training set

### ISSUE THREAD

**Issue by CaiFeipeng** [CongWeilin/mtcnn-caffe#44]
Hi @CongWeilin, 

Thanks for your great codes. I am following your work to reproduce MTCNN.

I have got some problems during training PNet and RNet.

1. I have trained a PNet on the randomly crpped images, but when the trained caffemodel was applied on WIDER FACE training set to generate RNet training data, the recall rate was quite low, and therefore, only 1k positive faces were obtained, resulting in many unrecognized faces on testing images. (Notice that I simply kept the threshold as 0.6.)

2. Dose the question 1 caused by the improper threshold? How did you identify the threadshold for each Net. Did you set them according to the racall rate on FDDB?

3. I used the false positive samples generated from PNet plus all true positive samples(around 200k) as RNet training data, but the trained RNet caffemodel did not work. It refused almost all true faces in several test images.

All the best.

**Comment by walkerwjt** [comment]
@CaiFeipeng hi, i have the similiar quesion with you, i have trained a PNet, but when i used the trained model to generate hard examples for RNet, too many false-positive examples were obtained, but my training acc is 89%-95%, sometimes it can reach 100%. Have you solved your problem?

i have compared the gen_12net_data.py with Seanlinx' gen_pnet_data.py, and i have found some diffrences:

1. size = npr.randint(40, min(width, height) / 2), in Seanlinx' gen_pnet_data.py, the num is 12, it seems that the original size of samples is more likely to be bigger,

2. the process of generating the negative samples is also different,

@CongWeilin , will the diffrences affect my training process and detection effect?

All the best.

**Comment by CaiFeipeng** [comment]
@walkerwjt No, I did not solve the issue. The project has been suspended.

As for the differences of the two scripts, I don't think they are the main reasons that result in massive false positive examples at testing time.

If you come up with any method to address the problem, please let me know. Thanks!

**Comment by AITTSMD** [comment]
@walkerwjt @CaiFeipeng  I meet the same problem !!! the p-net model I trained is very bad but I can get 92% accuracy rate in validation data. the r-net model is also bad!!! When I generate hard examples for r-net using p-net model  offered by the author, I get 9W+ positive samples.

**Comment by walkerwjt** [comment]
@CaiFeipeng @AITTSMD, i have trained a P-net model before, and when i used it to generate hard examples for R-net, it took a long time. The recall rate was also quite low. I changed "caffe_img = img.copy()-128" to "caffe_img = (img.copy()-127.5)/127.5" in gen_12net_hard_wxample.py, then the process of generating hard examples was much faster. But the recall rate is still low. I only get 18k positive samples, but there are more than two million false-positive samples. 

And i have tried to train a R-net model by using the data, the pefermance is also bad. When i used the trained P-net and R-net model to generate hard examples for O-net, i found that the network can detect rectangles only when I set the threshold to [0.5, 0.5]. I also found that the part face image is processed by "im -= 128" while others are "im = (im - 127.5)/127.5"  in "gen_net_imdb.py ". I don't understand why the handling is different here. Except that, i haven't found any solution yet.

If i have maked any progress, i will share. And if you have any ideas, please let me know. Thanks, all the best!

**Comment by LucyLu-LX** [comment]
@walkerwjt  Same problem's here. 

I trained Pnet on WiderFace for cls and roi. The accuracy of Pnet is 90%-95% on training set, and on  validation it reaches 90%. The loss of roi is around 0.03 and doesn't change too much. But it generates too many rectangles, which takes much longer time in processing than it should. I have to increase the threshold for Pnet. I don't know if it is reasonable to do so.

![873701eb1cb0dfdd2d312db3efafccb](https://user-images.githubusercontent.com/23068004/28307326-1e4af360-6bd5-11e7-9c2c-cd3db3509c10.png)

![9321a95bd1debee9e066f9058a3b065](https://user-images.githubusercontent.com/23068004/28307333-23b4f3d2-6bd5-11e7-9ee6-957edb9f5cf2.png)


But for Rnet, it is the opposite, I need to decrease the threshold, otherwise it wouldn't generate rectangles at all.

I changed "im -= 128" to "im = (im - 127.5)/127.5" in "gen_net_imdb.py ", otherwise, it will cause the loss for roi is nan for my training process.

**Comment by CaiFeipeng** [comment]
@walkerwjt  @LucyLu-LX It should be a mistake. I changed im -= 128 to im = (im -127.5)/127.5. I also tried many other training tricks, like [this one](https://github.com/foreverYoungGitHub/MTCNN), but nothing helped. That's why I suspended the project.

**Comment by AITTSMD** [comment]
@LucyLu-LX Hi,I wonder how many positive samples dou u get using pnet to generate data for training rnet!

**Comment by LucyLu-LX** [comment]
@AITTSMD I use widerface for training. Cause my pnet generates too many rectangles, looks like it would take forever to generate hard samples. So I limit the output to 500. In this way, I get 9k positive samples.

**Comment by walkerwjt** [comment]
Hi, I have made some progress. You can also try this way, or maybe you have better solutions.

    1.  I changed im -= 128 to im = (im -127.5)/127.5 in "gen_net_imdb.py ";
    2.  Referring to the process of generating data in Seanlinx' "gen_pnet_data.py", I changed "size = npr.randint(40, min(width, height) / 2)" to "size = npr.randint(12, min(width, height) / 2)" which means that the size of samples is more likely to be smaller. (But I didn't changed the process of generating the negative samples. You can modify it, and maybe the result will be better.)

In the way, I have trained a P-net model. And when I used it to generate hard examples, I have get 17k+    positive samples while the progress bar is 33%. It seems that I will get 51k+ positive samples eventually. And I have spent 38 hours so far while the threshold is set to 0.7. Hopefully you have better solutions.

**Comment by walkerwjt** [comment]
Hi, I have made some progress. You can also try this way, or maybe you have better solutions.

    1.  I changed im -= 128 to im = (im -127.5)/127.5 in "gen_net_imdb.py ";
    2.  Referring to the process of generating data in Seanlinx' "gen_pnet_data.py", I changed "size =
 npr.randint(40, min(width, height) / 2)" to "size = npr.randint(12, min(width, height) / 2)" which means
 that the size of samples is more likely to be smaller. (But I didn't changed the process of generating the
 negative samples. You can modify it, and maybe the result will be better.)

In the way, I have trained a P-net model. And when I used it to generate hard examples, I have get 17k+    positive samples while the progress bar is 33%. It seems that I will get 50k+ positive samples eventually. And I have spent 38 hours so far while the threshold is set to 0.7. Hopefully you have better solutions.

**Comment by walkerwjt** [comment]
Hi, I have made some progress. You can also try this way, or maybe you have better solutions.

1.  I changed im -= 128 to im = (im -127.5)/127.5 in "gen_net_imdb.py ";
2.  Referring to the process of generating data in Seanlinx' "gen_pnet_data.py", I changed "size = npr.randint(40, min(width, height) / 2)" to "size = npr.randint(12, min(width, height) / 2)" which means that the size of samples is more likely to be smaller. (But I didn't changed the process of generating the negative samples. You can modify it, and maybe the result will be better.)

In the way, I have trained a P-net model. And when I used it to generate hard examples, I have get 17k+    positive samples while the progress bar is 33%. It seems that I will get 50k+ positive samples eventually. And I have spent 38 hours so far while the threshold is set to 0.7. Hopefully you have better solutions.

**Comment by AITTSMD** [comment]
@LucyLu-LX @walkerwjt @CaiFeipeng HI,I reproduced mtcnn using tensorflow.The result is okay.I can get 14W+ pos for training RNet and 19W+ pos for training ONet.Please infer https://github.com/AITTSMD/MTCNN-Tensorflow!

**Comment by walkerwjt** [comment]
@AITTSMD  Wow, that's great! Thank you for your sharing, I will study it in detail. By the way, did you find the reason why it has failed all the time on Caffe? Since my project is mainly based on Caffe, can you share what changes you have done, especially about the data processing part and training part? Many thanks!

**Comment by AITTSMD** [comment]
@walkerwjt hi, I mainly change the training method.I construct a batch that contains pos,part,neg and landmark data.Use different datas to conduct different tasks.That means,I conduct 3 tasks at the same time rather than one by one.

**Comment by CaiFeipeng** [comment]
@AITTSMD  I also performed 3 tasks in each minibatch using Caffe, but the experimental results were not as good as I expected. How did you make it?  BTW, thanks for your great work!

**Comment by AITTSMD** [comment]
@CaiFeipeng May the loss function used in Caffe has something wrong.You can try the experiment in tensorflow.

**Comment by CaiFeipeng** [comment]
@AITTSMD  It is possible. I am running the tf scripts now. Thanks for your work!

**Comment by AITTSMD** [comment]
@CaiFeipeng you are welcome,you can  propose  the issue if have any problems.

**Comment by yungs2017** [comment]
@LucyLu-LX How you train the Pnet to get that result? when i training the Pnet, i always get the cls loss 0.693147
